recipes-openxt/openxt/openxt-measuredlaunch/ml-functions

```diff
@@ -0,0 +1,375 @@
+#!/bin/sh
+#
+# Copyright (c) 2016 Daniel P. Smith, Apertus Solutions, LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+#
+
+. /usr/lib/openxt/key-functions
+[ $? -eq 0 ] || {
+    echo "failed to load key-functions"
+    exit 1
+}
+
+hex2bin() {
+    read hex
+    echo -n $hex | sed 's/\([0-9A-F]\{2\}\)/\\\\\\x\1/gI' | xargs printf
+}
+
+# Function to parse grub.cfg and return all multiboot modules and their
+#   full command line
+# parameter 0: path to grub.conf
+parse_grub() {
```

Comment:
See [Phil's writeup](http://twobit.us/2013/06/calculating-pcr18-and-pcr19/) on PCRs 18 and 19 to understand what goes into the measurement. Other items about txt-stat is it does not tell you the filepath for the module it measured and it maps /dev/mem which is a poor security practice. The addition of a module is not the only change that would affect the measurement. Others include a module's commandline args changing or the order they are listed could change. The measurements are sensitive to all of these changes, any of which could happen during an upgrade. The whole point of this work is to predict the measurements of what will be not what currently is, which is what txt-stat is giving you.